### PR TITLE
Dicebox

### PR DESCRIPTION
**Pull request by SniderThanYou**
This PR adds a new generator for a dice box. It is based on the `IntegratedHingeBox`, but requires some extra finger joints for the dice separator panels. I added the finger joint support to three of the `ChestHingeXYZ` edges. If there's a better way to do it, please let me know.

I also updated `regularPolygonHole()` to support rounded corners, because it just looks a little nicer than the sharp corners. I used the `FillTest` generator to verify that the existing functionality is preserved. `FillTest` uses `fillHoles()`, which uses `regularPolygonHole()`

PXL_20220810_021014702

### CODE REVIEW

**Comment by florianfesti** [comment]
This looks very nice.

But I am currently on vacation. So reviews can take longer or shorter than normal.

**Comment by SniderThanYou** [comment]
@florianfesti take your time :) There is no need to rush.

File: boxes/edges.py
Comment (florianfesti) [inline boxes/edges.py:1470]: These need to be absolute_params. There is really no need to scale the bool with thickness

File: boxes/edges.py
Comment (florianfesti) [inline boxes/edges.py:1464]: Please make this two separate lines as they are used to generate the descriptions in the UI.

For now there is no way to have hidden parameters that ar enot shown in the UI but are set only by the generators. But this is not the first place where such thing is useful and I might add that feature at some point in the future. Probably somehthing like parameters starting with `_` being hidden.

As they for now can be deselected by the user you can either not render the wall with the hex holes for the side that got unchecked or may be force the setting to `True` in the generator code.

File: boxes/edges.py
Comment (florianfesti) [inline boxes/edges.py:1520]: I wonder if it is easier to read of you just use a length for the finger joint edge and set that to `0.0` for the normal case. But that's more a question of style.

File: boxes/generators/dicebox.py
Comment (florianfesti) [inline boxes/generators/dicebox.py:96]: Mixing `y` and `x` wide parts in the first column is not ideal. This also breaks if y is smaller than x. Either rotate the parts or render them in the right `x` wide column.

File: boxes/generators/dicebox.py
Comment (florianfesti) [inline boxes/generators/dicebox.py:58]: I wonder if this can be done more easier and more generically by using either the new [fillHoles](https://github.com/florianfesti/boxes/blob/master/boxes/__init__.py#L1558) function or one of the old [hex functions](https://github.com/florianfesti/boxes/blob/master/boxes/__init__.py#L1896) probably [hexHolesHex](https://github.com/florianfesti/boxes/blob/master/boxes/__init__.py#L1987).

File: boxes/__init__.py
Comment (florianfesti) [inline boxes/__init__.py:1256]: I am not 100% decide what I think of that. From an API POV the default radius should be a true `0.0` including the over cuts to make things fit. Otoh these holes are mostly not used for constructive purposes. May we can make the default radius `None` and translate that to `burn` but allow to pass `0.0` for getting the over cuts. This still doesn't play well wit making the radius an UI parameter, tough. May be I am overthinking this...

**Review by florianfesti — CHANGES_REQUESTED** [review]
This looks pretty good already code vise. The few thing that really need changing should be very easy to fix.

**Comment by SniderThanYou** [comment]
@florianfesti Thanks for the review, Florian! I'm looking through your comments now.

File: boxes/__init__.py
Comment (SniderThanYou) [reply boxes/__init__.py:1256]: @florianfesti It looks like the original author of `regularPolygonalHole` intended this:

`if the radius is less than the burn, then make sure we at least cut _something_, instead of leaving the surface solid`.

However, this seems like a VERY rare edge case, especially since `regularPolygonalHole` is only used by `fillHoles`, and now the dice box. That said, I don't really mind either way. For the purposes of the dice box, no one will ever set the radius to zero; I'm happy to set the radius however you wish. But perhaps that adjustment could be a separate commit, since related to the original implementation, instead of the addition of rounded corners.

The corner radius, on the other hand, could absolutely be zero, with no issue. At the very least, the `+ 1E-9` should be removed from the corner radius burn adjustment. I'll make this change now.

File: boxes/edges.py
Comment (SniderThanYou) [reply boxes/edges.py:1470]: Thanks, fixed

File: boxes/edges.py
Comment (SniderThanYou) [reply boxes/edges.py:1464]: Thanks, fixed

File: boxes/generators/dicebox.py
Comment (SniderThanYou) [reply boxes/generators/dicebox.py:96]: Good catch! The defaults create a square box, so I didn't notice it when testing. Fixed.

File: boxes/edges.py
Comment (SniderThanYou) [reply boxes/edges.py:1527]: @florianfesti I'll leave this one up to you; this is your code base, and it's important for you to be happy with the style. So if that's how you would prefer, then I'd be happy to make the change.

File: boxes/generators/dicebox.py
Comment (SniderThanYou) [reply boxes/generators/dicebox.py:58]: @florianfesti It looks like `hexHolesHex` creates circular holes, in a hexagon pattern, in the shape of a hexagon (rather than filling an entire rectangle with holes). But `fillHoles` could do the job more generically. I don't think that it would be easier, though. I need to get to bed, but I can take a look again tomorrow. Thanks again for all the great suggestions!

Comment (SniderThanYou) [reply boxes/generators/dicebox.py:64]: @florianfesti I experimented with `fillHoles()`, and encountered the following problems:

- the distance between the hexagons horizontally is not the same as the distance between them vertically, so it's not as visually appealing
- there's no way to force the radius of the hexagons to be exactly what we want; there's too much internal manipulation of the values
- it really wants to start by putting a hexagon in the top left corner of the area, thus creating a `3-2-3` arrangement, instead of the `2-3-2` arrangement common to dice boxes
- if you add a new parameter to pass through `corner_radius`, then that parameter doesn't make sense for `round` holes, or for scaled-down holes

For now, it seems better to use `regularPolygonalHole()` directly. This was a good experiment though, because it made me realize that my code also had irregular spacing between the x and y directions. I fixed it by just converting to polar coordinates; it was so much easier, and much clearer what we wanted to accomplish.

**Comment by SniderThanYou** [comment]
@florianfesti Back to you, friend! This is ready for a second review.

File: boxes/generators/dicebox.py
Comment (florianfesti) [inline boxes/generators/dicebox.py:115]: This one is still the wrong way round and only moves y to the side when it needs to offset the x column.

Comment (SniderThanYou) [reply boxes/generators/dicebox.py:116]: @florianfesti ohhh I get it; I did not really understand the "only" moves before. Fixed, thanks.

Comment (florianfesti) [reply boxes/generators/dicebox.py:116]: Yeah, it isn't that obvious but very convenient once one understands how to use it...

File: boxes/generators/dicebox.py
Comment (florianfesti) [reply boxes/generators/dicebox.py:64]: Yeah, I guess `fillHoles` - which is really new - needs a bit of polishing to be usable here. It currently lacks the means to align the pattern symmetrically with borders and to control which horizontal offset to start with the hex pattern. That on my todo list as it is necessary to replace the older `hex*` functions 

In the long run I can imagine it being used to have an arbitrary number of holes depending on the size of the box. Just hard coding it to 7 holes is fine for now, though.

File: boxes/edges.py
Comment (florianfesti) [reply boxes/edges.py:1527]: Fair enough. Let's not hold this up on questions of style.

**Comment by florianfesti** [comment]
With your permission I would cut the picture to size and add it to the generator. If you want to do that yourself there are brief instructions at https://florianfesti.github.io/boxes/html/CONTRIBUTING.html#provide-photos-for-generators.

**Comment by florianfesti** [comment]
There's an argument to be made that the generator should calculate the size of the hex holes according to size. But if we plan to fit the proper amount of holes at some point, there is no good reason to drop it now.

**Comment by florianfesti** [comment]
Anyway thanks for your work!

**Comment by SniderThanYou** [comment]
> With your permission I would cut the picture to size and add it to the generator. If you want to do that yourself there are brief instructions at https://florianfesti.github.io/boxes/html/CONTRIBUTING.html#provide-photos-for-generators.

Please be my guest. I release the image to the public domain and waive any rights I hold over it.

> There's an argument to be made that the generator should calculate the size of the hex holes according to size. But if we plan to fit the proper amount of holes at some point, there is no good reason to drop it now.

Doesn't it already?

```
apothem = (min(xi, yi) - 4 * t) / 6
r = apothem * 2 / math.sqrt(3)
```

> Anyway thanks for your work!

My pleasure! Thanks for sharing the tool.

**Comment by HaSHsss** [comment]
it's looks like not OK with this fingerhole

image

**Comment by SniderThanYou** [comment]
> it's looks like not OK with this fingerhole
> 
> image

Those tabs fit into the integrated hinge:

PXL_20220818_184628816

PXL_20220818_184639063

**Comment by florianfesti** [comment]
Well, this generator - and the original IntegratedHingeBox - would benefit from labels for the parts.

**Pull request merged**
